### DISCUSSION TITLE
Where to contribute improved docblocks?

### DISCUSSION THREAD

**Discussion by xwillq** [larastan/larastan#1353] [Q&A] [answered]
Hi!
I want to improve some docblocks. For example, pagination contracts and paginators themselves could use some generics to specify, what data they paginate. Also kernel and middleware docblocks could be improved, so running Larastan on freshly created project would produce less errors.

The question is, where should I do it? Laravel itself includes generics and nice docblocks for some classes. Should I make pull request to Laravel repository, or just add/change Larastan stubs?

**Comment by canvural** [comment, marked as answer]
Hi,

> Larastan on freshly created project would produce less errors.

As far as I know there is no error on a fresh Laravel project on level 9 even. What errors do you see?

> The question is, where should I do it? Laravel itself includes generics and nice docblocks for some classes. Should I make pull request to  Laravel repository, or just add/change Larastan stubs?

Laravel only accepts generics on classes if it's gonna improve the auto-completion in IDEs also. So if your changes are about just to improve static analysis, you should propose them here. You can use the stub files here and add some tests. But if you think your changes would also improve the auto-completion in the IDEs you can try to submit them to Laravel itself. And if it gets rejected you can always submit them here later.

**Reply by xwillq** [reply]
My bad. If you don't count `Undefined variable: $this` in routes/console.php and errors in tests, then everything is fine. All the errors were found because of `phpstan/phpstan-strict-rules`.

Is there any point in adding stubs for properties in kernel, middleware and providers? If yes, I will try to add them some time later to fix some of the errors from `phpstan/phpstan-strict-rules`.

**Reply by canvural** [reply]
Yeah, anything that makes the types more specific would welcome 👍🏽

**Comment by mad-briller** [comment]
@xwillq good news, generic paginators are on their way! https://github.com/nunomaduro/larastan/pull/1372
